### DISCUSSION TITLE
How can resolve "error: invalid conversion from ‘void*’ to ‘long int*’ [-fpermissive]"

### DISCUSSION THREAD

**Discussion by FelisCatuus** [community/community#102763] [Programming Help]
### Body

#include <stdio.h>
#include <stdlib.h>

int main()
{
    int a;
    scanf("%d", &a);
    long int* b = malloc(sizeof(long) * (a + 1));

    for (int i = 0; i < a; i++)
    {
        int c, d;
        scanf("%d %d", &c, &d);
        b[i] = c + d;
    }
    for (int i = 0; i < a; i++)
        printf("%ld\n", b[i]);

    return 0;
}

In line 8 the error was occured.
When I run this code in visual studio with C++, the code run well.
But I run this in test cite, the code can't run with the error.
What is a problem and how can I solve this?

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by gil-son** [comment]
The issue in your code may be related to the type mismatch in the malloc call. You are allocating memory for long int elements, but in line 8, you are trying to assign int values to the array. This can lead to undefined behavior and might cause issues, especially on certain online judges or test sites.

Try to do that:

```
#include <stdio.h>
#include <stdlib.h>

int main() {
    int a;
    scanf("%d", &a);

    long int* b = malloc(sizeof(long int) * a);

    for (int i = 0; i < a; i++) {
        int c, d;
        scanf("%d %d", &c, &d);
        b[i] = (long int)c + d;
    }

    for (int i = 0; i < a; i++)
        printf("%ld\n", b[i]);

    free(b); 

    return 0;
}
```

**Comment by offensive-vk** [comment]
## Few Points to be Noticed: 
- malloc returns void*: The malloc function allocates memory and returns a generic pointer of type void*.
- Direct assignment to long int*: You're trying to assign the result of malloc (which is void*) directly to a pointer of type long int*.
- Type mismatch: void* can point to any type of data, so the compiler doesn't know what kind of data b actually points to when assigning it directly.
* Instead try this:
```c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

int main() {
    int i, a;
	
    printf("Enter the number of elements: ");
    if (scanf("%d", &a) != 1 || a <= 0) {
        fprintf(stderr, "Invalid input: Please enter a positive integer.\n");
        return 1;
    }

    if (a > INT_MAX / sizeof(long int)) {
        fprintf(stderr, "Error: Cannot allocate memory for this many elements.\n");
        return 1;
    }

    long int* b = (long int*)malloc(sizeof(long int) * (a + 1));
    if (b == NULL) {
        fprintf(stderr, "Memory allocation failed.\n");
        return 1;
    }

    printf("Enter pairs of numbers (c d):\n");
    for (i = 0; i < a; i++) {
        int c, d;
        if (scanf("%d %d", &c, &d) != 2) {
            fprintf(stderr, "Error reading input for element %d.\n", i + 1);
            free(b); 
            return 1;
        }
        b[i] = c + d;
    }


    printf("Results:\n");
    for (i = 0; i < a; i++) {
        printf("%ld\n", b[i]);
    }

    free(b);
    return 0;
}

```
